### PR TITLE
Configure RNGs appropriately for Pipeline + SPMD

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #689

DTensor has existing RNG management. It requires a shared seed for every
rank in its 'world' (SPMD world).  Then it manages offsets per rank
using its own RNG tracker to ensure same or different random values
across ranks depending on the device-mesh and the type of sharding on
the current operation being performed.  (TODO: link to docs)

When used together with pipeline parallelism, it is important to use a
different seed for each separate SPMD world.  E.g. if the user specified
seed 1234, then we can literally use 1234 for all the ranks on PP=0, but
then we should use a different seed (e.g. 1234 + 1) for ranks on PP=1.
This partitions the world into PP separate SPMD worlds and uses a unique
seed for each SPMD world.

Control 'deterministic' mode separately from rng seed

The use case for 'deterministic' mode may be more for debugging, while
users may want to control RNG seeds used for real runs.

### CODE REVIEW

File: torchtitan/utils.py [hunk 1]
```diff
@@ -52,6 +52,34 @@ def _warn_overwrite_env(env, val):
...
+
+    Currently, does not set seeds for the CUDA RNG since TorchTitan always uses DTensor for SPMD parallelisms,
+    and DTensor manages its own RNG tracker, but we could extend to support both if needed.
+    """
+    # to ensure we can control which ranks have same or different seeds, all ranks agree on a starting seed.
+    # if user provides one, we use this. Otherwise rank 0 rolls the dice and everyone else uses that.
+    if seed is None:
+        # Extract the seed for torch's main generator on rank 0 and standardizes on using that to build
+        # seeds for unique SPMD groups
+        seed = torch.get_rng_state()[:8].view(torch.uint64).item()
```
Comment (fegin) [inline torchtitan/utils.py]: Why `[:8]`?

File: torchtitan/utils.py [hunk 2]
```diff
@@ -52,6 +52,34 @@ def _warn_overwrite_env(env, val):
...
+        seed_tensor = torch.tensor(seed, device=device, dtype=torch.int64)
+        torch.distributed.broadcast(seed_tensor, src=0)
+        seed = seed_tensor.item()
```
Comment (fegin) [inline torchtitan/utils.py]: Should we worry that users may give different manual seed on different ranks for some reasons (bugs?)?

File: torchtitan/utils.py [hunk 3]
```diff
@@ -52,6 +52,34 @@ def _warn_overwrite_env(env, val):
...
+
+    if parallel_dims.pp_enabled:
+        seed += parallel_dims.pp_rank
+
+    torch.distributed.tensor._random.manual_seed(seed)
```
Comment (tianyu-l) [inline torchtitan/utils.py]: `manual_seed` would need the spmd mesh as arg

File: torchtitan/utils.py (hunk 2 above)
Comment (wconstab) [reply torchtitan/utils.py]: i didn't worry about that, it seems like if you are passing a --seed argument from some launch thing, it would be a lot harder to 'accidentally' get different seeds per rank.  

for the case the user did not pass anything, i think each rank will start with its own seed. so for that case i think we have no choice but to broadcast.

File: torchtitan/utils.py (hunk 1 above)
Comment (wconstab) [reply torchtitan/utils.py]: iiuc the state tensor has [seed(8), offset(8)] together as 128 bits, but its represented as 16 elements of 'byte' tensor.  Actually i'm not sure if this is documented well or even reliable.

Comment (wconstab) [reply torchtitan/utils.py]: what i really want to do here is just 'get a good initial seed'.  I could use numpy or something.  but i felt i didn't want to reinvent the wheel.  open to suggestions

File: torchtitan/utils.py (hunk 3 above)
Comment (wconstab) [reply torchtitan/utils.py]: i think it doesn't, since xilun's latest PR just skips the broadcast when you give it a seed.  he is very trusting now, the caller will do the right thing :)

Comment (tianyu-l) [reply torchtitan/utils.py]: I saw it's still needed here https://github.com/pytorch/pytorch/pull/141223/files#diff-1c1c89e657502ab6c7e901dc740ce8f20c4226399b1982c37b4f153d5c8fa5dbR50

It seems it's for initializing the `OffsetBasedRNGTracker` correctly?

Comment (wconstab) [reply torchtitan/utils.py]: oh, my bad. ok, adding it back.

**Review by tianyu-l — COMMENTED** [review]
Looks good to me. Could you rebase onto https://github.com/pytorch/torchtitan/pull/597 to incorporate the latest changes there?

File: train.py [hunk 4]
```diff
@@ -81,6 +72,14 @@ def main(job_config: JobConfig):
     if parallel_dims.pp_enabled:
         pp_mesh = world_mesh["pp"]
 
+    # Set random seed, and maybe enable deterministic mode (mainly for debugging, expect perf loss)
+    utils.set_determinism(
+        spmd_mesh=world_mesh["dp", "tp"],
```
Comment (XilunWu) [inline train.py]: note that this only works for 3D (PP+DP+TP). Either "dp" or "tp" being absent causes error.

**Comment by XilunWu** [comment]
Successfully run test `CONFIG_FILE="./train_configs/debug_model.toml" NGPU=8 LOG_RANK=0,1,2,3,4,5,6,7 ./run_llama_train.sh` after modifying `pipeline_parallel_degree = 2` and `tensor_parallel_degree = 2` with pytorch PR https://github.com/pytorch/pytorch/pull/141532

File: train.py (hunk 4 above)
Comment (wconstab) [reply train.py]: is it safe to assume that every non-PP dimension is part of the 'spmd world'? 

if so, should we make a device-mesh api that lets you get the world minus pp?

Comment (XilunWu) [reply train.py]: I think this assumption is right. Though I'm not sure if this still holds for the EP coming. It can be directly added in `train.py` because this is likely the only place this logic gets used. Besides, `DeviceMesh` should not be aware of parallelisms IMO.

Comment (fegin) [reply train.py]: Yes, EP should still be SPMD world and the design of `DeviceMesh` is that it can be used without parallelisms. So it is better to have a util outside of `DeviceMesh`.

Comment (wconstab) [reply train.py]: yea, agree that device mesh shouldn't be aware of parellelisms.  Just wondered if device mesh should provide a helper API for filtering dimensions. Maybe not worth adding.  This is the code I wrote in train.py- and I wish it were a little cleaner:

`spmd_mesh_dims = list(filter(lambda name: name != "pp", world_mesh.mesh_dim_names))`

(note: even though train.py "knows" which parallelisms are used, its not that straightforward to write logic to construct a list of appropriate dims for spmd_mesh dims and get them in the right order.  I felt its easier/safer to just "filter out" the PP one.  Not sure if others agree.

**Review by XilunWu — APPROVED** [review]
LGTM! Thanks for the work getting torchtitan RNG in good shape!

It's worth note that in case of no DTensor being used and user doesn’t call `set_determinism` with a seed (i.e. specify `[training] seed` in job’s `.toml` file), then the random state is set by torch init.

File: torchtitan/utils.py [hunk 5]
```diff
@@ -52,26 +53,75 @@ def _warn_overwrite_env(env, val):
...
+        # https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
+    torch.backends.cudnn.deterministic = deterministic
+    torch.backends.cudnn.benchmark = not deterministic
+
+    # to ensure we can control which ranks have same or different seeds, all ranks agree on a starting seed.
+    # if user provides one, we use this. Otherwise rank 0 rolls the dice and everyone else uses that.
     if seed is not None:
         # CPU and GPU determinism
         torch.manual_seed(seed)
```
Comment (tianyu-l) [inline torchtitan/utils.py]: Could you remind me of the logic here? It seems right now:
1. every rank would call `torch.manual_seed` on the same seed, regardless of PP existence / ranks
2. different PP ranks would call `torch.distributed.tensor._random.manual_seed` on different seeds, on the spmd mesh.

What does `torch.manual_seed` control in torchtitan?

File: torchtitan/utils.py [hunk 6]
```diff
@@ -52,26 +53,75 @@ def _warn_overwrite_env(env, val):
...
+        )
+        spmd_mesh = world_mesh[spmd_mesh_dims] if len(spmd_mesh_dims) else None
+    else:
+        spmd_mesh = world_mesh
+        logger.debug(f"Global Rank {c10d.get_rank()} using seed: {seed}")
+
+    # As long as we are not in the 1-D (PP-only) case, we will have a seed to use for all ranks of the SPMD mesh.
+    # IF PP is also used, this seed is unique per PP rank.
+    if spmd_mesh:
+        torch.distributed.tensor._random.manual_seed(seed, spmd_mesh)
```
Comment (wconstab) [inline torchtitan/utils.py:114]: oops, i think i need to call torch.manual_seed()  again here, at least in the non-spmd case- to ensure the updated pp seed is used.

Comment (tianyu-l) [reply torchtitan/utils.py:114]: how about when `spmd_mesh is None`, do different PP ranks get different `torch.manual_seed(seed)`?

Comment (XilunWu) [reply torchtitan/utils.py:114]: or call `torch.manual_seed` no matter if `spmd` exists? It doesn't hurt in SPMD case since we use DTensor.

Comment (wconstab) [reply torchtitan/utils.py:114]: yea, i have made it so that manual_seed (and python seed, fwiw) are called no matter what.

File: train.py [hunk 7]
```diff
@@ -206,14 +199,7 @@ def loss_fn(pred, labels):
...
 
-    checkpoint_loaded = checkpoint.load()
-
-    if parallel_dims.pp_enabled and not checkpoint_loaded:
-        # TODO: fix this by allowing each rank to set their own seed
-        logger.warning(
-            "Pipeline Parallelism is being used without a seed checkpoint. "
-            "All the substages will be initialized with random weights with same RNG state which can affect convergence."
-        )
+    checkpoint.load()
```
Comment (tianyu-l) [inline train.py]: pls rebase to the change in https://github.com/pytorch/torchtitan/pull/716

**Review by tianyu-l — APPROVED** [review]
looks good to me. Could you take a look at the CI failure? other regular / PR tests are also failing due to that.

File: torchtitan/config_manager.py [hunk 8]
```diff
@@ -390,7 +390,13 @@ def __init__(self):
             "--training.seed",
             type=int,
             default=None,
-            help="Implement reproducibility by setting a Python, PyTorch and CUDA seed",
+            help="Choose the base RNG seed used for training",
+        )
+        self.parser.add_argument(
+            "--training.deterministic",
+            type=bool,
+            default=False,
```
Comment (carmocca) [inline torchtitan/config_manager.py]: If you follow the pattern in the rest of the args, this should be action=store_true
